starlark/profile.go

```
@@ -0,0 +1,419 @@
+	enc.bytes(Profile_period_type, wallNanos.Bytes()) // dimension and unit of period
+
+	start := nanotime()
+
+	// Read profile events from the channel
+	// until it is closed by StopProfiler.
+	for e := range profiler.events {
+		sample := new(bytes.Buffer)
+		sampleenc := protoEncoder{w: sample}
+		sampleenc.int(Sample_value, e.time.Nanoseconds()) // wall nanoseconds
+		for _, fr := range e.stack {
+			sampleenc.uint(Sample_location_id, location(fr))
+		}
+		enc.bytes(Profile_sample, sample.Bytes())
+	}
+
+	end := nanotime()
+	enc.int(Profile_time_nanos, start)         // start time of profile
+	enc.int(Profile_duration_nanos, end-start) // duration of profile
+
+	profiler.done <- gz.Close() // Close reports any prior write error
+}
+
+// profFuncAddr returns the canonical "address"
+// of a Callable for use by the profiler.
+func profFuncAddr(fn Callable) uintptr {
+	switch fn := fn.(type) {
+	case *Builtin:
+		return reflect.ValueOf(fn.fn).Pointer()
+	case *Function:
+		return uintptr(unsafe.Pointer(fn.funcode))
+	}
+
+	// User-defined callable types are typically of
+	// of kind pointer-to-struct. Handle them specially.
+	if v := reflect.ValueOf(fn); v.Type().Kind() == reflect.Ptr {
+		return v.Pointer()
+	}
+
+	// Address zero is reserved by the protocol.
+	// Use 1 for callables we don't recognize.
+	log.Printf("Starlark profiler: no address for Callable %T", fn)
+	return 1
+}
+
+// We encode the protocol message by hand to avoid making
+// the interpreter depend on both github.com/google/pprof
+// and github.com/golang/protobuf.
+//
+// This also avoids the need to materialize a protocol message object
+// tree of unbounded size and serialize it all at the end.
+// The pprof format appears to have been designed to
+// permit streaming implementations such as this one.
+//
+// See https://developers.google.com/protocol-buffers/docs/encoding.
+type protoEncoder struct {
+	w   io.Writer // *bytes.Buffer or *gzip.Writer
```

I don't think it's safe to ignore errors if you allow `*gzip.Writer` here. I'm not sure how much buffering `*gzip.Writer` has internally, but if the underlying `io.Writer` is unbuffered, this could be making a lot of system calls.

Maybe instead, accumulate a number of profile events in a `*bytes.Buffer`, then flush that when it reaches some threshold (maybe somewhere in the range of 8-64 KB)? You would only need to check for errors on flush.

---

Oops, I neglected your second question. The gzip.Writer compresses data in 240 byte chunks, so that's the size of each underlying write to the file descriptor. I added a bufio.Writer to write in 4KB chunks, though I suspect the speed improvement is intangible in practice.